### ISSUE TITLE
Inconsistencies among release content and https://github.com/oleg-shilo/cs-script/wiki/Choosing-Compiler-Engine

### ISSUE THREAD

**Issue by maettu-this** [oleg-shilo/cs-script#215]
https://github.com/oleg-shilo/cs-script/wiki/Choosing-Compiler-Engine

#### Mixup among C# 6 and 7
"C# 6 syntax for scripts is not available out of box. It is hard to believe but Microsoft did not released C# 7+ compiler with .NET 4.6. Though the work around is relatively simple. You just need to provide the path to the custom compiler for C# 6."

#### Not finding a "CSSCodeProvider.v4.6.dll" in the latest full release [3.30.3.0](https://github.com/oleg-shilo/cs-script/releases/tag/v3.30.3.0)
"If you have CS-Script present then it is only a single line of code:"
```
    CSScript.GlobalSettings.UseAlternativeCompiler = 
                        "%CSSCRIPT_DIR%\Lib\CSSCodeProvider.v4.6.dll";
```
But in neither `cs-script.7z` nor `cs-script.ExtensionPack.7z` I can find such a .dll.

**Comment by oleg-shilo** [comment]
Correct. `CSSCodeProvider.v4.6.dll` has been superseded with C# version neutral "CSSRoslynProvider.dll" which in turn invokes Roslyn from the configurable location. See https://github.com/oleg-shilo/cs-script/wiki/C%23-7-support-%28hosted%29

["Choosing-Compiler-Engine"](https://github.com/oleg-shilo/cs-script/wiki/Choosing-Compiler-Engine) has been updated to reflect that.

While CSSRoslynProvider.dll is available from distro, it is recommended to import the packages ("cs-script.bin" and "Microsoft.Net.Compilers") from the NuGet server.

**Comment by maettu-this** [comment]
OK, better now. But I find a few things still confusing:

- https://github.com/oleg-shilo/cs-script/wiki/Choosing-Compiler-Engine lists "C# 6 support | yes" in the table but later refers to a subsite talking about bringing C# 7 support.
- Section [CodeDom compiler] refering to a Roslyn compiler seems contradicting. In the end, I as a user don't see whether this will totally switch to Roslyn (bringing C# 7 for dumping `css_import` and the like) or whether I will still have CodeDom-like behavior with a Roslyn compiler (what seems another contradiction). I think the answer to this question should be stated there.
- C# 7 or 7.1 or 7.2 or 7.3? Yet another .NET versioning difficulty...
- https://github.com/oleg-shilo/cs-script/wiki/C%23-7-support-%28hosted%29 tells how to activate Roslyn. The first line of code `CSScript.GlobalSettings.UseAlternativeCompiler = ...` is self explaining. But why do I also need the second `CSScript.GlobalSettings.RoslynDir = ...`? This seems what we call "doppel-moppel" in German, something like double-glued, i.e. one of both seems unnecessary. A comment stating the reason would help.

Some more confusing things about `CSSRoslynProvider.dll`

- The assembly description of ".NET v4.6 (C#6) code complilation with Roslyn" seems outdated.
- The assembly version of "3.28.3.0" apparently refers to an earlier version of CS-Script. The date of something in 2018 seems to confirm this.
- `CSSCodeProvider` is namespace-less. What is the reason for this?


Having added the necessary things to my hosting application results in an exception:
`System.IO.FileLoadException: Die Datei oder Assembly "CSSRoslynProvider, Version=3.28.3.0, Culture=neutral, PublicKeyToken=null" oder eine Abhängigkeit davon wurde nicht gefunden. Eine Assembly mit starkem Namen ist erforderlich. (Ausnahme von HRESULT: 0x80131044)` meaning something like "**strong name required**", which `CSSRoslynProvider` doesn't have.

So this way to bring C# 6 or even C# 7 to my hosting application doesn't work, at least not that straight-forward.

**Comment by maettu-this** [comment]
I can live with the limitation of C# 5 for a while. Some users would like to see 6 or 7, but I can just tell them that this is quite new, there are technical restrictions and they have to wait for a future version.

And, referring to your answer to my recent email, it is likely that the near future brings more improvements in .NET 5. Just a few years ago, WinForms and WPF seemed far way for .NET Core, now they are in .NET 5 (though still limited to Windows, thus not yet replacing Mono). (I know Mono is dead, but there simply is no other choice for WinForms on Linux yet.) Maybe Microsoft will eventually realize that such functionality is indeed needed in the real world and will eventually fill the gaps.

**It is quite impressive reading how you provided features with CS-Script many years earlier than Microsoft eventually did in .NET! As I have depicted in my email, CS-Script is really powerful and useful for the hundreds of test scripts I work with. Please keep up this superb work!**

**Comment by oleg-shilo** [comment]
I am not sure about your problems with C# 5+ but if you follow the example given in the wiki you will probably succeed. At least with the current versions of the packages.

I have attached the example for you, just in case.
But ensure you resolved all packages when you build.
[ConsoleApp16.zip](https://github.com/oleg-shilo/cs-script/files/5533559/ConsoleApp16.zip)

**Comment by maettu-this** [comment]
First responding to the [Nov 12] post:

https://github.com/oleg-shilo/cs-script/wiki/C%23-7-support-%28hosted%29 states:
> you need add a special NuGet package (provided by Microsoft) to your project - "Microsoft.Net.Compilers".

This will work fine if the hosting application is executed on the developer's machine, i.e. the machine where the NuGet package is available for sure. But, in my case, where the users write their own scripts, allowing the users of the hosting application to make use of C#7 support will mean they have to install NuGet and get the NuGet package on their machine as well. Technically possible, but two more hurdles to be taken. A solution where no additional dependency on the user machine is required would be preferred, i.e. a solution where all assemblies can be deployed along with the hosting application.

**Comment by maettu-this** [comment]
Then reminding of my more recent "OK, better now. But..." post:

I have again read through my post. I think the statements are still valid, i.e. there is still quite some potential in the referred to wiki pages.

**Comment by oleg-shilo** [comment]
> A solution where no additional dependency on the user machine is required would be preferred, i.e. a solution where all assemblies can be deployed along with the hosting application.

This is not really a feasible solution as it would require CS-Script to be rereleased every time a new version of Microsoft.Net.Compilers package is released. 

Though you as a developer, as a vendor of your product, have full freedom of building a distribution packages that include any possible combination of CS-Script and MS.Net.Compilers.   

> .... I think the statements are still valid

Some are. But some are not :)

> lists "C# 6 support | yes"  

I have replaced all C# versions with "C# 6 (and above)" 

> C# 7 or 7.1 or 7.2 or 7.3?

Not sure what exact content you are referring to. But basically it is "C# 6 (and above)"

> But why do I also need the second ...

First line is telling that you want to switch to one of the three possible compiler engines.
The next line tells cs-script where to find your selected compiler because its location is not deterministic. 

>The assembly description of ".NET v4.6 (C#6) code complilation with Roslyn" seems outdated.

Will need to update next time the assembly is rebuild. 
But this version of the assembly was build to support C#6 (available at that time) and it just happens to support any other version above 6. Even though these versions were not available at the time of this assembly was built.

>The assembly version of "3.28.3.0" apparently refers to an earlier version of CS-Script. The date of something in 2018 seems to confirm this.

Correct. There were no changes to the assembly since then.

> CSSCodeProvider is namespace-less. What is the reason for this?

I would put it differently. What is the reason to use namespace in this case.

This assembly is a dynamically loaded plugin (code provider). The entrypoint in this assembly is located via Reflection thus all plugin implementations (when they come from various assemblies) must have this entry point with 100% identical full name. Thus it is completely opposite paradigm that namespaces are used for.

Namespeces are used to avoid naming collisions. When using same type names from multiple referenced assemblies.
CS-Script always uses only a single plugin assembly and it is not referenecd but dynamically loaded via Reflection so it needs to have a simple predictable name. Meaning that using namespace does not bring any value at all.

>  meaning something like "strong name required", which CSSRoslynProvider doesn't have.

This is because you are trying to "statically" reference CSSRoslynProvider.dll while its intended use is to be loaded dynamically specifically via Reflection. And dynamic loading does not utilize assembly signing (strong name) mechanism at all. So no signing. 

> So this way to bring C# 6 or even C# 7 to my hosting application doesn't work, at least not that straight-forward.

The code sample I provided (ConsoleApp16.zip) shows how to achieve that.

> Section [CodeDom compiler] refering

It is a difficult one to address without having a dedicated discussion. Let's have it in another thread or offline. I am sure there are some confusing bits but I am not confident how much effort should I put into content that is about to become obsolete.

As you know I am working on bringing CodeDom to .NET 5 build of CS-Script. This will bring the end to the compilers gigantic confusion. Why? Because .NET Core has all compilers being deployed with the default distribution as it was prior C#6 (2015). 
For me it is one of the strongest drivers for the migration to .NET 5.

**Comment by maettu-this** [comment]
Thanks for this extra bunch of information. Getting the compilers back in .NET 5 indeed looks like a big plus.

Still, as discussed before, people and applications need time to migrate. I think it is worth getting rid of potentially confusing content. After all, it looks like a thing that can be done within an hour or so.

By the way, propably your recent change has broken the wiki list of contents. The C# 7 hosted link now is https://github.com/oleg-shilo/cs-script/wiki/C%23-7-support-%28hosted%29#-7-support-(hosted) and no longer properly working (leading to home instead of C# 7).

With "C# 7 or 7.1 or 7.2 or 7.3?" I meant what is documented at https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7. For whatever reason, MS decided to increment C# 7 three times. So I guess when you talk about C# 7 you actually mean C# 7.3. Though arguable, stating C# 7 can be seen as covering the whole lot.

Tomorrow I'll try to "intended use is to be loaded dynamically specifically via Reflection". This very piece of information I didn't get from the wiki, thus added the assembly to my Visual Studio project like others. It would be really cool bringing C# 7 support to my hosting application now. I'll let you know once succeeded.

**Comment by maettu-this** [comment]
Alright, here's the result:

It's not feasible to ship the binaries along with my hosting app, I would have to ship the whole content of the `%USERPROFILE%\.nuget\packages\Microsoft.Net.Compilers\2.2.0\tools\` folder: `csc.exe`, more than 30 `.dll`,... This doesn't make much sense.

I could do something like `if (Microsoft.Net.Compilers NuGet package is available) then use alternative compiler; else standard` in my application. Still, users would have to get that NuGet package by themselves. This complicates use cases where a whole team uses a script, all team members would have to install the correct NuGet package 2.2.0, incl. on build and test servers. And the version is another pitfall:

The latest version of Microsoft.Net.Compilers is 3.8.0. But that states `Note: This package is deprecated. Please use Microsoft.Net.Compilers.Toolset instead`. Tried it anyway, resulting in my VS2017 project didn't build anymore:
```
1>------ Rebuild All started: Project: ConsoleApp16, Configuration: Debug Any CPU ------
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```
Even worse, I managed to "kill" my project, i.e. attempting uninstalling failed as well:
```
Attempting to gather dependency information for package 'Microsoft.Net.Compilers.2.2.0' with respect to project 'ConsoleApp16', targeting '.NETFramework,Version=v4.7.2'
Resolving actions to uninstall package 'Microsoft.Net.Compilers.2.2.0'
Unable to uninstall 'Microsoft.Net.Compilers.2.2.0' because 'CS-Script.RoslynProvider.3.30.3' depends on it.
========== Finished ==========
```
Two strange things:
- Error message above says `2.2.0` even though I upgraded the project to the `3.8.0` version.
- Error message above says `CS-Script.RoslynProvider.3.30.3` above while `CSSRoslynProvider` states version `3.28.3.0`.

I also tried to replace the `Console.WriteLine(...)` by `System.Windows.Forms.MessageBox.Show(...)`. This results in an exception. Adding the `System.Windows.Forms` to `ConsoleApp16` didn't help, neither to change from console to a windows app.

While all these hurdles could possible be taken, this seems just a too shaky approach for an application that is intended to run 24/7 x-thousand times a night on servers as well as local machines of experienced and unexperienced developers and technical staff. I'll wait with C# 6+ until having migrated to .NET 5.

**Comment by maettu-this** [comment]
I saw a few things got further refined and fixed in the wiki. I'll close this issue now, it's up to you on how to make use of my additional findings listed above.

**Comment by oleg-shilo** [comment]
as for "CS-Script.RoslynProvider.3.30.3" and all other Roslyn packaging struggle I am happy to say that with CS-Script for .NET5 it will no longer be a problem as .NET5 comes with all compilers available as it was always the case before C#6. 

And when hosting on .NET Framework app I am building a new csc shell `build.dll` similar to MS `VSCSCompiler.exe`. Though `build.dll` is using a different assembly probind/loading approach so it will always preload csc.exe so there will be no need for any additional proxy like CS-Script.RoslynProvider.3.30.3.

CS-Script .NET5 is particularly good for future engine hosting apps at the class lib is implemented as netstandard so the same package works in all environments.

You can start looking at it in the alpha branch: https://github.com/oleg-shilo/cs-script.core/tree/NET5_preview
